extensions/sql/asset/loader/src/main/java/org/eclipse/dataspaceconnector/sql/asset/loader/SqlAssetLoader.java

```diff
@@ -0,0 +1,307 @@
+        Objects.requireNonNull(dataAddress);
+
+        try (var connection = getConnection()) {
+            if (existsById(asset.getId(), connection)) {
+                throw new EdcPersistenceException(String.format("Cannot persist. Asset with ID '%s' already exists.", asset.getId()));
+            }
+
+            transactionContext.execute(() -> {
+                try {
+                    executeQuery(connection, SQL_ASSET_INSERT_CLAUSE_TEMPLATE,
+                            asset.getId());
+                    executeQuery(connection, SQL_DATA_ADDRESS_INSERT_CLAUSE_TEMPLATE,
+                            asset.getId(),
+                            objectMapper.writeValueAsString(dataAddress.getProperties()));
+
+                    for (var property : asset.getProperties().entrySet()) {
+                        executeQuery(connection, SQL_PROPERTY_INSERT_CLAUSE_TEMPLATE,
+                                asset.getId(),
+                                property.getKey(),
+                                objectMapper.writeValueAsString(property.getValue()),
+                                property.getValue().getClass().getName());
+                    }
+
+                } catch (JsonProcessingException e) {
+                    throw new EdcPersistenceException(e);
+                }
+
+            });
+
+        } catch (Exception e) {
+            if (e instanceof EdcPersistenceException) {
+                throw (EdcPersistenceException) e;
+            }
+            throw new EdcPersistenceException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void accept(AssetEntry item) {
+        Objects.requireNonNull(item);
+
+        this.accept(item.getAsset(), item.getDataAddress());
+    }
+
+    @Override
+    public Asset deleteById(String assetId) {
+        Objects.requireNonNull(assetId);
+
+        var asset = findById(assetId);
+        if (asset == null) {
+            return null;
+        }
+
+        try (var connection = getConnection()) {
+            transactionContext.execute(() -> {
+                executeQuery(connection, SQL_ASSET_PROPERTY_DELETE_BY_ID_CLAUSE_TEMPLATE, assetId);
+                executeQuery(connection, SQL_DATA_ADDRESS_DELETE_BY_ID_CLAUSE_TEMPLATE, assetId);
+                executeQuery(connection, SQL_ASSET_DELETE_BY_ID_CLAUSE_TEMPLATE, assetId);
+            });
+        } catch (Exception e) {
+            throw new EdcPersistenceException(e.getMessage(), e);
+        }
+
+        return asset;
+    }
+
+    @Override
+    public Stream<Asset> queryAssets(AssetSelectorExpression expression) {
+        Objects.requireNonNull(expression);
+
+        return getAssetStream(SQL_ASSET_LIST_CLAUSE_TEMPLATE);
+    }
+
+    @Override
+    public Stream<Asset> queryAssets(QuerySpec querySpec) {
+        Objects.requireNonNull(querySpec);
+
+        var limit = Limit.Builder.newInstance().limit(querySpec.getLimit()).offset(querySpec.getOffset()).build();
```

Comment:
the `Limit` seems awfully complicated for managing two ints, both of which have sane defaults in the `QuerySpec`. Consequently, the `LIMIT` and `OFFSET` clause could be coded directly into the select statement.